src/latency.c

```diff
@@ -490,6 +490,89 @@ sds createLatencyReport(void) {
+ * Empty buckets are not printed.
+ * Everything above 1 sec is considered +Inf.
+ * At max there will be log2(1000000000)=30 buckets */
+void fillCommandCDF(client *c, struct hdr_histogram* histogram){
+    addReplyMapLen(c,2);
+    addReplyBulkCString(c,"calls");
+    addReplyLongLong(c,(long long) histogram->total_count);
+    addReplyBulkCString(c,"histogram_usec");
+    void *replylen = addReplyDeferredLen(c);
+    int samples = 0;
+    struct hdr_iter iter;
+    hdr_iter_log_init(&iter,histogram,1024,2);
+    int64_t previous_count = 0;
+    while (hdr_iter_next(&iter)) {
+        const int64_t micros = iter.highest_equivalent_value / 1000;
+        const int64_t cumulative_count = iter.cumulative_count;
+        if(cumulative_count > previous_count){
+            addReplyLongLong(c,(long long) micros);
+            addReplyLongLong(c,(long long) cumulative_count);
+            samples++;
+        }
+        previous_count = cumulative_count;
+    }
+    setDeferredMapLen(c,replylen,samples);
+}
+
+/* latencyCommand() helper to produce for all commands,
+ * a per command cumulative distribution of latencies. */
+void latencyAllCommandsFillCDF(client *c) {
+    dictIterator *di = dictGetSafeIterator(server.commands);
+    dictEntry *de;
+    struct redisCommand *cmd;
+    void *replylen = addReplyDeferredLen(c);
+    int command_with_data = 0;
+    while((de = dictNext(di)) != NULL) {
+        char *tmpsafe;
+        cmd = (struct redisCommand *) dictGetVal(de);
+        if (!cmd->latency_histogram)
+            continue;
+        const char* cmdname = getSafeInfoString(cmd->name, strlen(cmd->name), &tmpsafe);
+        addReplyBulkCString(c,cmdname);
+        fillCommandCDF(c, cmd->latency_histogram);
+        if (tmpsafe != NULL) zfree(tmpsafe);
+        command_with_data++;
+    }
+    dictReleaseIterator(di);
+    setDeferredMapLen(c,replylen,command_with_data);
+}
+
+/* latencyCommand() helper to produce for a specific command set,
+ * a per command cumulative distribution of latencies. */
+void latencySpecificCommandsFillCDF(client *c) {
+    void *replylen = addReplyDeferredLen(c);
+    int valid_commands = 0;
+    for (int j = 2; j < c->argc; j++){
+        struct redisCommand *cmd = dictFetchValue(server.commands, c->argv[j]->ptr);
+        /* If the command does not exist we skip the reply */
+        if (cmd == NULL) {
+            continue;
+        }
+        char *tmpsafe;
+        const char* cmdname = getSafeInfoString(c->argv[j]->ptr, strlen(c->argv[j]->ptr), &tmpsafe);
+        addReplyBulkCString(c,cmdname);
+        valid_commands++;
+        /* If no latency info we reply with the same format as non empty histograms */
+        if (!cmd->latency_histogram) {
+            addReplyMapLen(c,2);
+            addReplyBulkCString(c,"calls");
+            addReplyLongLong(c,0);
+            addReplyBulkCString(c,"histogram_usec");
+            addReplyMapLen(c,0);
+            continue;
+        }
```

Comment:
i don't like this, the outer function is aware of the response layout of the inner one.
if we must have that reply, we should call `fillCommandCDF` and let it handle that internally.

however, i think it's excessive to add so much data for the reply, and i think that instead we should let the client realize the request came empty handed.
since we don't rely on the ordering anyway, i think that's perfectly valid approach.

Reply:
> however, i think it's excessive to add so much data for the reply, and i think that instead we should let the client realize the request came empty handed.

addressed in latest commit :+1: